File: javascript/src/1005.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
function largestSumAfterKNegations(nums, k) {
  nums.sort((a, b) => Math.abs(a) - Math.abs(b))
  for (let i = nums.length - 1; i >= 0 && k; i--) {
    if (nums[i] < 0) {
      nums[i] *= -1
      k--
    }
  }
  if (k & 1) {
    nums[0] *= -1
  }
  return nums.reduce((a, b) => a + b)
}

{
  const { strictEqual } = require('assert')

  const CASES = [
    [[4, 2, 3], 1, 5],
    [[3, -1, 0, 2], 3, 6],
    [[2, -3, -1, 5, -4], 2, 13],
  ]

  for (const [nums, k, excepted] of CASES) {
    const actual = largestSumAfterKNegations(nums, k)
    strictEqual(actual, excepted)
  }

  console.time()
  for (const [nums, k, _] of CASES) {
    for (let i = 0; i < 100000; i++) {
      largestSumAfterKNegations(nums, k)
    }
  }
  console.timeEnd()
}

File: python/src/1005.py
class Solution:

    def largestSumAfterKNegations(self, nums: list[int], k: int) -> int:
        nums.sort(key=abs)
        for i in range(len(nums) - 1, -1, -1):
            if not k:
                break
            if nums[i] < 0:
                nums[i] *= -1
                k -= 1
        if k & 1:
            nums[0] *= -1
        return sum(nums)


if __name__ == '__main__':
    import unittest
    from timeit import timeit

    CASES = (
        ([4, 2, 3], 1, 5),
        ([3, -1, 0, 2], 3, 6),
        ([2, -3, -1, 5, -4], 2, 13),
    )

    class Test(unittest.TestCase):

        def test(self):
            largestSumAfterKNegations = Solution().largestSumAfterKNegations
            for nums, k, excepted in CASES:
                actual = largestSumAfterKNegations(nums, k)
                self.assertEqual(actual, excepted)

    unittest.main()

    largestSumAfterKNegations = Solution().largestSumAfterKNegations
    for nums, k, _ in CASES:
        print(timeit(lambda: largestSumAfterKNegations(nums, k), number=100000))
